# 0.0 Problem Statement

Predict the sold price of the property that sold on 2020

# 1.1 Take a Quick Look at Data Structure

The goal of this section is to analyze the data and only keep data that is likely to be useful to us. Processing data for ML algorithm will be done in later sections

We have the housing data as before. But now we are only interested in listings that sold in 2020. We will also add additional features such as Interest Rate, and assessement value of the properties from 2020 and 2019. See the get_data variable in PreprocessingPipelineV2 for details on the data we will use for our purpose

In [50]:
import PreprocessingPipelineV2

In [51]:
house_data_dir = r"C:\Users\KI PC\OneDrive\Documents\Software Engineering and Computer Science\Internships\Riipen - KnockNow\BC-House-Pricing-Model"
seasons_data_path = r"C:\Users\KI PC\OneDrive\Documents\Software Engineering and Computer Science\Internships\Riipen - KnockNow\BC-House-Pricing-Model\month_seasons.csv"
#mortgage rate from here: https://www.ratehub.ca/historical-mortgage-rates-widget
interest_rate_2020_path = r"C:\Users\KI PC\OneDrive\Documents\Software Engineering and Computer Science\Internships\Riipen - KnockNow\BC-House-Pricing-Model\2020_month_by_month_interest_rate.csv"
assesement_data_path = r"C:\Users\KI PC\OneDrive\Documents\Software Engineering and Computer Science\Internships\Riipen - KnockNow\BC-House-Pricing-Model\West-van-assessments.csv"
longitude_latitude_data_path = r"C:\Users\KI PC\OneDrive\Documents\Software Engineering and Computer Science\Internships\Riipen - KnockNow\BC-House-Pricing-Model\longitude_latitude_data.csv"

In [52]:
#get data
import pandas as pd
house_data = PreprocessingPipelineV2.load_data(house_data_dir, "Spreadsheet")
assesement_data = pd.read_csv(assesement_data_path)
seasons_data = pd.read_csv(seasons_data_path)
interest_rate_2020_data = pd.read_csv(interest_rate_2020_path)

In [53]:
house_data_sold = PreprocessingPipelineV2.get_data.fit_transform([house_data, assesement_data, seasons_data, interest_rate_2020_data])

C:\Users\KI PC\OneDrive\Documents\Software Engineering and Computer Science\Internships\Riipen - KnockNow\BC-House-Pricing-Model\PreprocessingPipelineV2.py:296: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  X[col_name] = X[col_name].str.replace(bad_char, "")


In [54]:
house_data_sold.head()

Address  DOM  Tot BR  Tot Baths TotFlArea   Age  \
73     2675 SKILIFT PLACE   16       3          2     3,182  62.0   
74     2675 SKILIFT PLACE   16       3          2     3,182  62.0   
332       1575 TAYLOR WAY    8       4          2     2,100  62.0   
360  4278 ROCKRIDGE PLACE   34       4          4     5,561  38.0   
400      2045 27TH STREET   29       4          3     2,307  54.0   

    Frontage - Feet  #Kitchens  List Date  Sold Date  ...  \
73              0.0          1   4/7/2020  4/23/2020  ...   
74              0.0          1   4/7/2020  4/23/2020  ...   
332          105.00          2  4/20/2020  4/28/2020  ...   
360          129.00          1   3/3/2020   4/6/2020  ...   
400           74.61          1   3/9/2020   4/7/2020  ...   

     2020 Buildings Value  2019 Total Value  2019 Land Value  \
73                55900.0         1794900.0        1739000.0   
74                55900.0         1794900.0        1739000.0   
332               67600.0         1343400.0        1277000.0   
360             1212000.0         3107000.0        1896000.0   
400              243000.0         1590000.0        1462000.0   

     2019 Buildings Value  Interest Rate  Season    S/A TypeDwel  \
73                55900.0           2.39  Spring  VWVCP    HOUSE   
74                55900.0           2.39  Spring  VWVCP    HOUSE   
332               66400.0           2.39  Spring  VWVBP    HOUSE   
360             1211000.0           2.39  Spring  VWVRR    HOUSE   
400              128000.0           2.39  Spring  VWVQU    HOUSE   

        Showing Appts Sold Price  
73   Phone L.R. First    1441110  
74   Phone L.R. First    1441110  
332  Phone L.R. First    1555000  
360  Phone L.R. First    3300000  
400       Phone Other    1900000  

[5 rows x 22 columns]

We have the following columns:

In [55]:
house_data_sold.columns

Index(['Address', 'DOM', 'Tot BR', 'Tot Baths', 'TotFlArea', 'Age',
       'Frontage - Feet', '#Kitchens', 'List Date', 'Sold Date',
       '2020 Total Value', '2020 Land Value', '2020 Buildings Value',
       '2019 Total Value', '2019 Land Value', '2019 Buildings Value',
       'Interest Rate', 'Season', 'S/A', 'TypeDwel', 'Showing Appts',
       'Sold Price'],
      dtype='object')

We have seen them in PricingEstimateV1. The new ones are described below

2019/2020 Total Value, 2019/2020 Land Value, 2019/2020 Buildings Value: These are the assessed value of the properties from https://www.bcassessment.ca/

Season: Fall / Summer / Winter / Spring

Interest Rate: 5 year Mortgage Rate in the month that the property sold (https://www.ratehub.ca/historical-mortgage-rates-widget)

## 1.2 Train and Test Split

We will again use stratified sampling





In [56]:
strat_train, strat_train_labels, strat_test, strat_test_labels =  PreprocessingPipelineV2.create_train_test_set(df = house_data_sold)

In [57]:
print(strat_train.shape)
print(strat_train_labels.shape)
print(strat_test.shape)
print(strat_test_labels.shape)

(424, 21)
(424,)
(106, 21)
(106,)


# 2.0 Prepare the Data for Machine Learning Algorithms

See the pipeline data_processing_pipeline in PreprocessingPipelineV2 for details on how the data was transformed for machine learning algorithms. Most of them stayed the same from PreprocessingPipeline

# 3.0 Selecting, Fine Tuning and Training a Model

As of now, we have two data sets: Training Set and Test Set. Typically training set is further divided into a training set and a validation set. The reduced training set is used to train the model, and the validation set is used to fine-tune the model. For our purpose, we will use repeated cross-validation instead. In repeated cross-validation, we use many small validation sets. Each model is evaluated once per validation set after it is trained on the rest of the data. By averaging out all the evaluations of a model, we get a much accurate measure of its performance. The drawback is that the training time is multiplied by the number of validation sets.

We will use the following models for testing: Decision Tree Regressor, Random Forest Regressor, and SVR

## 3.1 Initial Model(s) Selection

First, we need to preprocess the data

In [58]:
strat_train_prepared = PreprocessingPipelineV2.data_preprocessing_pipeline.fit_transform(strat_train)

In [59]:
strat_train_prepared.shape

(424, 53)

The columns are as follows

In [60]:
attributes = PreprocessingPipelineV2.numerical_columns + PreprocessingPipelineV2.ordinal_columns + PreprocessingPipelineV2.one_hot_encoding_cols_catgs

In [61]:
len(attributes)

53

In [62]:
from sklearn.model_selection import cross_val_score

In [63]:
#Function to show cross validation scores
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Standard deviation:", scores.std())

### 3.1.1 Decision Tree Regressor

In [64]:
from sklearn.tree import DecisionTreeRegressor
tree_reg = DecisionTreeRegressor(random_state=42)
tree_reg.fit(strat_train_prepared, strat_train_labels)

DecisionTreeRegressor(random_state=42)

In [65]:
import numpy as np
from sklearn.model_selection import cross_val_score

scores_tree = cross_val_score(tree_reg, strat_train_prepared, strat_train_labels, scoring="neg_mean_squared_error", cv=10)
tree_rmse_scores = np.sqrt(-scores_tree)
display_scores(tree_rmse_scores)

Scores: [ 496943.08687704  636420.60127524  914666.88929473 1448290.38186877
  698541.10598351  429317.21992432  374942.58980327  654613.23848848
  457112.13375203  910746.89990439]
Mean: 702159.4147171775
Standard deviation: 305884.05211487605


### 3.1.2 Random Forest Regressor

In [66]:
from sklearn.ensemble import RandomForestRegressor

forest_reg = RandomForestRegressor(n_estimators=100, random_state=42)
forest_reg.fit(strat_train_prepared, strat_train_labels)

RandomForestRegressor(random_state=42)

In [67]:
scores_forest = cross_val_score(forest_reg, strat_train_prepared, strat_train_labels, scoring="neg_mean_squared_error", cv=10)
forest_rmse_scores = np.sqrt(-scores_forest)
display_scores(forest_rmse_scores)

Scores: [ 318072.45277284  489482.46107972  779418.15622912 1371477.40167161
  602370.21004875  351979.76264135  362542.29479952  356409.2720032
  413793.86893444  584656.5714782 ]
Mean: 563020.2451658745
Standard deviation: 302909.37692627567


The results are still not great but The Random Forest Regressor is outperforming the Decision Tree Regressor by a signifciant margin

### 3.1.3 SVR

In [68]:
from sklearn.svm import SVR
svm_reg = SVR(kernel="linear")
svm_reg.fit(strat_train_prepared, strat_train_labels)

SVR(kernel='linear')

In [69]:
scores_svr = cross_val_score(svm_reg, strat_train_prepared, strat_train_labels, scoring="neg_mean_squared_error", cv=10)
svm_reg_rmse_scores = np.sqrt(-scores_svr)
display_scores(svm_reg_rmse_scores)

Scores: [1017245.98067136 1755003.78742336 2325827.45504266 1603963.69046168
 1556436.30768348 1586522.13425066 1763226.03033128 2165512.46990569
 1445170.70347792 2393613.48578087]
Mean: 1761252.2045028948
Standard deviation: 404075.22895487625


The Random Forest Regressor is outperfomring by quite a margin but we may get better results with SVR when we use Grid Search or Randomized Search. So for next stage, we will only keep SVR and Random Forest Regressor

In addition, Random Forest Algorithm combines the results of many decision trees to come up with results. Thus, it is unlikely to be helpful with ensemble learning. In addition, as shown above, it is likely to give better results than Decision Trees. So we will not consider Decision Tree any further.

# 3.2 Final Model Selection

### 3.2.1 Grid Search with Random Forest Regressor

In [70]:
from sklearn.model_selection import GridSearchCV

param_grid_forest = [
    {'n_estimators': [3, 10, 30], 'max_features': [2, 4, 6, 8]},
    {'bootstrap': [False], 'n_estimators': [3, 10], 'max_features': [2, 3, 4]},
  ]

forest_reg = RandomForestRegressor(random_state=42)
# train across 5 folds, that's a total of (12+6)*5=90 rounds of training 
grid_search_forest = GridSearchCV(forest_reg, param_grid_forest, cv=5,
                           scoring='neg_mean_squared_error',
                           return_train_score=True)
grid_search_forest.fit(strat_train_prepared, strat_train_labels)

GridSearchCV(cv=5, estimator=RandomForestRegressor(random_state=42),
             param_grid=[{'max_features': [2, 4, 6, 8],
                          'n_estimators': [3, 10, 30]},
                         {'bootstrap': [False], 'max_features': [2, 3, 4],
                          'n_estimators': [3, 10]}],
             return_train_score=True, scoring='neg_mean_squared_error')

In [71]:
print(grid_search_forest.best_params_)
print(grid_search_forest.best_estimator_)
print("Best Score: {0}".format(np.sqrt(-grid_search_forest.best_score_)))

{'max_features': 8, 'n_estimators': 30}
RandomForestRegressor(max_features=8, n_estimators=30, random_state=42)
Best Score: 601987.8302498491


In [72]:
pd.DataFrame(grid_search_forest.cv_results_).head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.006800      0.000748         0.001200        0.000402   
1       0.016800      0.000748         0.001999        0.000633   
2       0.044803      0.004706         0.004000        0.001549   
3       0.005600      0.000482         0.001000        0.000633   
4       0.015403      0.000488         0.001797        0.000399   

  param_max_features param_n_estimators param_bootstrap  \
0                  2                  3             NaN   
1                  2                 10             NaN   
2                  2                 30             NaN   
3                  4                  3             NaN   
4                  4                 10             NaN   

                                    params  split0_test_score  \
0   {'max_features': 2, 'n_estimators': 3}      -6.273827e+11   
1  {'max_features': 2, 'n_estimators': 10}      -4.837539e+11   
2  {'max_features': 2, 'n_estimators': 30}      -3.603174e+11   
3   {'max_features': 4, 'n_estimators': 3}      -5.934493e+11   
4  {'max_features': 4, 'n_estimators': 10}      -3.380196e+11   

   split1_test_score  ...  mean_test_score  std_test_score  rank_test_score  \
0      -1.004348e+12  ...    -6.167228e+11    2.047617e+11               15   
1      -8.589941e+11  ...    -5.695747e+11    1.728787e+11               12   
2      -7.309165e+11  ...    -4.662399e+11    1.725506e+11                6   
3      -1.317531e+12  ...    -8.336671e+11    2.914236e+11               18   
4      -1.005885e+12  ...    -5.809663e+11    2.351697e+11               13   

   split0_train_score  split1_train_score  split2_train_score  \
0       -1.903604e+11       -1.831241e+11       -1.700660e+11   
1       -9.550661e+10       -9.895396e+10       -1.245221e+11   
2       -8.681863e+10       -6.922355e+10       -8.028099e+10   
3       -2.389847e+11       -1.223119e+11       -1.986085e+11   
4       -1.067061e+11       -6.903301e+10       -1.220344e+11   

   split3_train_score  split4_train_score  mean_train_score  std_train_score  
0       -2.365833e+11       -2.030991e+11     -1.966466e+11     2.265163e+10  
1       -1.314448e+11       -1.088218e+11     -1.118498e+11     1.405079e+10  
2       -9.001358e+10       -8.630304e+10     -8.252796e+10     7.358694e+09  
3       -2.948977e+11       -1.492871e+11     -2.008180e+11     6.185661e+10  
4       -9.688951e+10       -9.078276e+10     -9.708917e+10     1.755967e+10  

[5 rows x 23 columns]

Better way to see the result is as following:

In [73]:
cvres = grid_search_forest.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

785317.0169891203 {'max_features': 2, 'n_estimators': 3}
754701.7100342035 {'max_features': 2, 'n_estimators': 10}
682817.5886134165 {'max_features': 2, 'n_estimators': 30}
913053.7179444979 {'max_features': 4, 'n_estimators': 3}
762211.4770261773 {'max_features': 4, 'n_estimators': 10}
673447.6275200925 {'max_features': 4, 'n_estimators': 30}
702450.1714624757 {'max_features': 6, 'n_estimators': 3}
628593.7846166254 {'max_features': 6, 'n_estimators': 10}
616840.6627070004 {'max_features': 6, 'n_estimators': 30}
703799.130474267 {'max_features': 8, 'n_estimators': 3}
639161.9986615975 {'max_features': 8, 'n_estimators': 10}
601987.8302498491 {'max_features': 8, 'n_estimators': 30}
698959.7281941043 {'bootstrap': False, 'max_features': 2, 'n_estimators': 3}
781769.1307630754 {'bootstrap': False, 'max_features': 2, 'n_estimators': 10}
795373.4184242416 {'bootstrap': False, 'max_features': 3, 'n_estimators': 3}
699233.5655407768 {'bootstrap': False, 'max_features': 3, 'n_estimators': 10}

### 3.2.2 Randomized Search with Random Forest Regressor

In [74]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint

param_distribs_forest = {
        'n_estimators': randint(low=1, high=200),
        'max_features': randint(low=1, high=8),
    }

forest_reg = RandomForestRegressor(random_state=42)
rnd_search_forest = RandomizedSearchCV(forest_reg, param_distributions=param_distribs_forest,
                                n_iter=50, cv=5, scoring='neg_mean_squared_error', random_state=42, verbose = 2)
rnd_search_forest.fit(strat_train_prepared, strat_train_labels)

Fitting 5 folds for each of 50 candidates, totalling 250 fits
[CV] END ...................max_features=7, n_estimators=180; total time=   0.2s
[CV] END ...................max_features=7, n_estimators=180; total time=   0.2s
[CV] END ...................max_features=7, n_estimators=180; total time=   0.2s
[CV] END ...................max_features=7, n_estimators=180; total time=   0.2s
[CV] END ...................max_features=7, n_estimators=180; total time=   0.2s
[CV] END ....................max_features=5, n_estimators=15; total time=   0.0s
[CV] END ....................max_features=5, n_estimators=15; total time=   0.0s
[CV] END ....................max_features=5, n_estimators=15; total time=   0.0s
[CV] END ....................max_features=5, n_estimators=15; total time=   0.0s
[CV] END ....................max_features=5, n_estimators=15; total time=   0.0s
[CV] END ....................max_features=3, n_estimators=72; total time=   0.0s
[CV] END ....................max_features=3, n_

[CV] END ....................max_features=6, n_estimators=51; total time=   0.0s
[CV] END ....................max_features=6, n_estimators=51; total time=   0.0s
[CV] END ....................max_features=6, n_estimators=51; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=55; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=55; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=55; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=55; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=55; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=64; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=64; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=64; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=64; total time=   0.0s
[CV] END ...................

[CV] END ....................max_features=4, n_estimators=54; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=54; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=54; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=54; total time=   0.0s
[CV] END ....................max_features=4, n_estimators=54; total time=   0.0s
[CV] END ...................max_features=5, n_estimators=191; total time=   0.2s
[CV] END ...................max_features=5, n_estimators=191; total time=   0.2s
[CV] END ...................max_features=5, n_estimators=191; total time=   0.2s
[CV] END ...................max_features=5, n_estimators=191; total time=   0.3s
[CV] END ...................max_features=5, n_estimators=191; total time=   0.2s
[CV] END ....................max_features=2, n_estimators=44; total time=   0.0s
[CV] END ....................max_features=2, n_estimators=44; total time=   0.0s
[CV] END ...................

RandomizedSearchCV(cv=5, estimator=RandomForestRegressor(random_state=42),
                   n_iter=50,
                   param_distributions={'max_features': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001E2FDC79548>,
                                        'n_estimators': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001E2FDC79C88>},
                   random_state=42, scoring='neg_mean_squared_error',
                   verbose=2)

In [75]:
print(rnd_search_forest.best_params_)
print(rnd_search_forest.best_estimator_)
print("Best Score: {0}".format(np.sqrt(-rnd_search_forest.best_score_)))

{'max_features': 7, 'n_estimators': 21}
RandomForestRegressor(max_features=7, n_estimators=21, random_state=42)
Best Score: 597118.7263403998


In [76]:
cvres = rnd_search_forest.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

610493.9513835558 {'max_features': 7, 'n_estimators': 180}
635846.5777701759 {'max_features': 5, 'n_estimators': 15}
660612.0421488424 {'max_features': 3, 'n_estimators': 72}
625845.665329867 {'max_features': 5, 'n_estimators': 21}
612291.9698166979 {'max_features': 7, 'n_estimators': 122}
658861.4752989073 {'max_features': 3, 'n_estimators': 75}
662880.8092722378 {'max_features': 3, 'n_estimators': 88}
617954.6412901207 {'max_features': 5, 'n_estimators': 100}
676708.6448309913 {'max_features': 3, 'n_estimators': 150}
821358.4860454007 {'max_features': 5, 'n_estimators': 2}
630477.0112112766 {'max_features': 4, 'n_estimators': 158}
608521.6767186042 {'max_features': 6, 'n_estimators': 130}
683867.454348174 {'max_features': 4, 'n_estimators': 21}
753624.5518343959 {'max_features': 1, 'n_estimators': 58}
610471.190043131 {'max_features': 6, 'n_estimators': 89}
752090.2838143584 {'max_features': 1, 'n_estimators': 59}
610883.8846923917 {'max_features': 7, 'n_estimators': 170}
633375.2712

In [77]:
feature_importances = rnd_search_forest.best_estimator_.feature_importances_
feature_importances

array([5.23431251e-03, 1.00603291e-02, 1.68633872e-02, 2.13808485e-02,
       6.76107968e-02, 6.38264885e-02, 2.49877518e-02, 9.21014169e-03,
       3.26564885e-03, 2.04495366e-01, 9.50977845e-02, 1.19120986e-01,
       9.48402398e-02, 1.32009193e-01, 7.24589401e-02, 1.42026009e-02,
       2.80186135e-03, 7.29832784e-05, 1.34630723e-04, 3.74522089e-03,
       1.69770121e-04, 2.04113363e-03, 8.93689189e-03, 2.99551829e-05,
       2.09308651e-04, 1.15190870e-03, 3.52623002e-03, 1.23741583e-04,
       5.04298572e-04, 1.25105258e-03, 1.09505390e-04, 5.48412427e-04,
       1.13768737e-04, 2.40634230e-03, 7.32983672e-04, 1.01596988e-04,
       8.45591087e-04, 3.52851870e-04, 6.38705169e-06, 1.51127286e-04,
       6.98000754e-05, 4.89402304e-03, 4.41842861e-04, 4.39502577e-05,
       6.39710321e-05, 2.01593904e-04, 8.77352422e-04, 4.57336877e-03,
       2.28891061e-03, 1.37489539e-03, 2.33119113e-04, 9.98174208e-07,
       2.03804318e-04])

In [78]:
sorted(zip(feature_importances, attributes), reverse = True)

[(0.2044953659872032, '2020 Total Value'),
 (0.13200919337561481, '2019 Land Value'),
 (0.11912098644625593, '2020 Buildings Value'),
 (0.0950977844539251, '2020 Land Value'),
 (0.09484023980010282, '2019 Total Value'),
 (0.07245894011662099, '2019 Buildings Value'),
 (0.06761079677858357, 'Tot Baths'),
 (0.06382648852754028, 'TotFlArea'),
 (0.024987751760874616, 'Age'),
 (0.02138084845121098, 'Tot BR'),
 (0.016863387249508748, 'DOM'),
 (0.014202600944366863, 'Interest Rate'),
 (0.010060329106629047, 'Longitude'),
 (0.009210141690632024, 'Frontage - Feet'),
 (0.008936891888287111, 'S/A: VWVCB'),
 (0.005234312512722461, 'Latitude'),
 (0.0048940230404315825, 'S/A: VWVWB'),
 (0.0045733687720541875, 'TypeDwel: HOUSE'),
 (0.003745220888916812, 'S/A: VWVBP'),
 (0.0035262300151536483, 'S/A: VWVCW'),
 (0.0032656488524369577, '#Kitchens'),
 (0.0028018613525735915, 'Season'),
 (0.0024063423030515653, 'S/A: VWVHB'),
 (0.002288910607650492, 'TypeDwel: MANUF'),
 (0.002041133630494995, 'S/A: VWVCA')

### 3.2.3 Grid Search with SVR

In [79]:
param_grid_svr = [
        {'kernel': ['linear'], 'C': [10., 30., 100., 300., 1000., 3000., 10000., 30000.0]},
        {'kernel': ['rbf'], 'C': [1.0, 3.0, 10., 30., 100., 300., 1000.0],
         'gamma': [0.01, 0.03, 0.1, 0.3, 1.0, 3.0]},
    ]
svm_reg = SVR()
grid_search_svr = GridSearchCV(svm_reg, param_grid_svr, cv=5,
                           scoring='neg_mean_squared_error',
                           verbose = 2) 

grid_search_svr.fit(strat_train_prepared, strat_train_labels)

Fitting 5 folds for each of 50 candidates, totalling 250 fits
[CV] END ..............................C=10.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=10.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=10.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=10.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=10.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=30.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=30.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=30.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=30.0, kernel=linear; total time=   0.0s
[CV] END ..............................C=30.0, kernel=linear; total time=   0.0s
[CV] END .............................C=100.0, kernel=linear; total time=   0.0s
[CV] END .............................C=100.0, 

[CV] END .....................C=10.0, gamma=0.01, kernel=rbf; total time=   0.0s
[CV] END .....................C=10.0, gamma=0.01, kernel=rbf; total time=   0.0s
[CV] END .....................C=10.0, gamma=0.01, kernel=rbf; total time=   0.0s
[CV] END .....................C=10.0, gamma=0.03, kernel=rbf; total time=   0.0s
[CV] END .....................C=10.0, gamma=0.03, kernel=rbf; total time=   0.0s
[CV] END .....................C=10.0, gamma=0.03, kernel=rbf; total time=   0.0s
[CV] END .....................C=10.0, gamma=0.03, kernel=rbf; total time=   0.0s
[CV] END .....................C=10.0, gamma=0.03, kernel=rbf; total time=   0.0s
[CV] END ......................C=10.0, gamma=0.1, kernel=rbf; total time=   0.0s
[CV] END ......................C=10.0, gamma=0.1, kernel=rbf; total time=   0.0s
[CV] END ......................C=10.0, gamma=0.1, kernel=rbf; total time=   0.0s
[CV] END ......................C=10.0, gamma=0.1, kernel=rbf; total time=   0.0s
[CV] END ...................

[CV] END .....................C=300.0, gamma=0.3, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=0.3, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=0.3, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=0.3, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=1.0, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=1.0, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=1.0, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=1.0, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=1.0, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=3.0, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=3.0, kernel=rbf; total time=   0.0s
[CV] END .....................C=300.0, gamma=3.0, kernel=rbf; total time=   0.0s
[CV] END ...................

GridSearchCV(cv=5, estimator=SVR(),
             param_grid=[{'C': [10.0, 30.0, 100.0, 300.0, 1000.0, 3000.0,
                                10000.0, 30000.0],
                          'kernel': ['linear']},
                         {'C': [1.0, 3.0, 10.0, 30.0, 100.0, 300.0, 1000.0],
                          'gamma': [0.01, 0.03, 0.1, 0.3, 1.0, 3.0],
                          'kernel': ['rbf']}],
             scoring='neg_mean_squared_error', verbose=2)

In [80]:
print(grid_search_svr.best_params_)
print(grid_search_svr.best_estimator_)
print("Best Score: {0}".format(np.sqrt(-grid_search_svr.best_score_)))

{'C': 30000.0, 'kernel': 'linear'}
SVR(C=30000.0, kernel='linear')
Best Score: 603296.7599865013


In [81]:
cvres = grid_search_svr.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

1793499.78408896 {'C': 10.0, 'kernel': 'linear'}
1765421.1956663218 {'C': 30.0, 'kernel': 'linear'}
1674471.4634736108 {'C': 100.0, 'kernel': 'linear'}
1418665.2338340327 {'C': 300.0, 'kernel': 'linear'}
902798.2349069767 {'C': 1000.0, 'kernel': 'linear'}
695040.4594470321 {'C': 3000.0, 'kernel': 'linear'}
625301.4135437836 {'C': 10000.0, 'kernel': 'linear'}
603296.7599865013 {'C': 30000.0, 'kernel': 'linear'}
1807776.4770630375 {'C': 1.0, 'gamma': 0.01, 'kernel': 'rbf'}
1807773.337576398 {'C': 1.0, 'gamma': 0.03, 'kernel': 'rbf'}
1807783.6837041117 {'C': 1.0, 'gamma': 0.1, 'kernel': 'rbf'}
1807790.8486750093 {'C': 1.0, 'gamma': 0.3, 'kernel': 'rbf'}
1807791.6115472065 {'C': 1.0, 'gamma': 1.0, 'kernel': 'rbf'}
1807791.6288915463 {'C': 1.0, 'gamma': 3.0, 'kernel': 'rbf'}
1807746.0986627382 {'C': 3.0, 'gamma': 0.01, 'kernel': 'rbf'}
1807736.680441969 {'C': 3.0, 'gamma': 0.03, 'kernel': 'rbf'}
1807767.708481518 {'C': 3.0, 'gamma': 0.1, 'kernel': 'rbf'}
1807789.22537678 {'C': 3.0, 'gamma':

### 3.2.4 Randomized Search with SVR

In [82]:
from scipy.stats import expon, reciprocal
param_distribs_svr = {
        'kernel': ['linear', 'rbf'],
        'C': reciprocal(20, 200000),
        'gamma': expon(scale=1.0),
    } 

svm_reg = SVR()
rnd_search_svr = RandomizedSearchCV(svm_reg, param_distributions=param_distribs_svr,
                                n_iter=50, cv=5, scoring='neg_mean_squared_error',
                                verbose=2, random_state=42)
rnd_search_svr.fit(strat_train_prepared, strat_train_labels)

Fitting 5 folds for each of 50 candidates, totalling 250 fits
[CV] END C=629.782329591372, gamma=3.010121430917521, kernel=linear; total time=   0.0s
[CV] END C=629.782329591372, gamma=3.010121430917521, kernel=linear; total time=   0.0s
[CV] END C=629.782329591372, gamma=3.010121430917521, kernel=linear; total time=   0.0s
[CV] END C=629.782329591372, gamma=3.010121430917521, kernel=linear; total time=   0.0s
[CV] END C=629.782329591372, gamma=3.010121430917521, kernel=linear; total time=   0.0s
[CV] END C=26290.206464300216, gamma=0.9084469696321253, kernel=rbf; total time=   0.0s
[CV] END C=26290.206464300216, gamma=0.9084469696321253, kernel=rbf; total time=   0.0s
[CV] END C=26290.206464300216, gamma=0.9084469696321253, kernel=rbf; total time=   0.0s
[CV] END C=26290.206464300216, gamma=0.9084469696321253, kernel=rbf; total time=   0.0s
[CV] END C=26290.206464300216, gamma=0.9084469696321253, kernel=rbf; total time=   0.0s
[CV] END C=84.14107900575871, gamma=0.059838768608680676, 

[CV] END C=8935.505635947808, gamma=0.37354658165762367, kernel=rbf; total time=   0.0s
[CV] END C=8935.505635947808, gamma=0.37354658165762367, kernel=rbf; total time=   0.0s
[CV] END C=8935.505635947808, gamma=0.37354658165762367, kernel=rbf; total time=   0.0s
[CV] END C=135.76775824842434, gamma=0.838636245624803, kernel=linear; total time=   0.0s
[CV] END C=135.76775824842434, gamma=0.838636245624803, kernel=linear; total time=   0.0s
[CV] END C=135.76775824842434, gamma=0.838636245624803, kernel=linear; total time=   0.0s
[CV] END C=135.76775824842434, gamma=0.838636245624803, kernel=linear; total time=   0.0s
[CV] END C=135.76775824842434, gamma=0.838636245624803, kernel=linear; total time=   0.0s
[CV] END C=151136.20282548846, gamma=1.4922453771381408, kernel=rbf; total time=   0.0s
[CV] END C=151136.20282548846, gamma=1.4922453771381408, kernel=rbf; total time=   0.0s
[CV] END C=151136.20282548846, gamma=1.4922453771381408, kernel=rbf; total time=   0.0s
[CV] END C=151136.2028

[CV] END C=15415.161544891856, gamma=0.2691677514619319, kernel=rbf; total time=   0.0s
[CV] END C=15415.161544891856, gamma=0.2691677514619319, kernel=rbf; total time=   0.0s
[CV] END C=15415.161544891856, gamma=0.2691677514619319, kernel=rbf; total time=   0.0s
[CV] END C=1888.9148509967113, gamma=0.739678838777267, kernel=linear; total time=   0.0s
[CV] END C=1888.9148509967113, gamma=0.739678838777267, kernel=linear; total time=   0.0s
[CV] END C=1888.9148509967113, gamma=0.739678838777267, kernel=linear; total time=   0.0s
[CV] END C=1888.9148509967113, gamma=0.739678838777267, kernel=linear; total time=   0.0s
[CV] END C=1888.9148509967113, gamma=0.739678838777267, kernel=linear; total time=   0.0s
[CV] END C=55.53838911232773, gamma=0.578634378499143, kernel=linear; total time=   0.0s
[CV] END C=55.53838911232773, gamma=0.578634378499143, kernel=linear; total time=   0.0s
[CV] END C=55.53838911232773, gamma=0.578634378499143, kernel=linear; total time=   0.0s
[CV] END C=55.53838

RandomizedSearchCV(cv=5, estimator=SVR(), n_iter=50,
                   param_distributions={'C': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001E28433BF48>,
                                        'gamma': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001E28433BA88>,
                                        'kernel': ['linear', 'rbf']},
                   random_state=42, scoring='neg_mean_squared_error',
                   verbose=2)

In [83]:
print(rnd_search_svr.best_params_)
print(rnd_search_svr.best_estimator_)
print("Best Score: {0}".format(np.sqrt(-rnd_search_svr.best_score_)))

{'C': 34246.75194632794, 'gamma': 0.3632878599687583, 'kernel': 'linear'}
SVR(C=34246.75194632794, gamma=0.3632878599687583, kernel='linear')
Best Score: 600956.8596322286


In [84]:
cvres = rnd_search_svr.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

1094765.5486976888 {'C': 629.782329591372, 'gamma': 3.010121430917521, 'kernel': 'linear'}
1807344.9496448098 {'C': 26290.206464300216, 'gamma': 0.9084469696321253, 'kernel': 'rbf'}
1806646.3432579075 {'C': 84.14107900575871, 'gamma': 0.059838768608680676, 'kernel': 'rbf'}
1270339.8344447585 {'C': 432.37884813148855, 'gamma': 0.15416196746656105, 'kernel': 'linear'}
1807790.7945334218 {'C': 24.17508294611391, 'gamma': 3.503557475158312, 'kernel': 'rbf'}
1586630.6881574679 {'C': 113564.03940586245, 'gamma': 0.0007790692366582295, 'kernel': 'rbf'}
1807738.7037246767 {'C': 108.30488238805073, 'gamma': 0.3627537294604771, 'kernel': 'rbf'}
1777854.636281951 {'C': 21.344953672647435, 'gamma': 0.023332523598323388, 'kernel': 'linear'}
1787297.7379854897 {'C': 5603.270317432516, 'gamma': 0.15023452872733867, 'kernel': 'rbf'}
1663547.001301633 {'C': 157055.10989448498, 'gamma': 0.26497040005002437, 'kernel': 'rbf'}
605066.231728336 {'C': 27652.464358739708, 'gamma': 0.2227358621286903, 'kernel'

The best model is the Random Forest Regressor model model obtaind from Randomized Search with mean error of Best Score: 597118.7263403998

In [85]:
final_model = rnd_search_forest.best_estimator_

## 3.3 Generalization Error and Confidence Interval For Final Model

I skipped caclualting error on test set as the error on validation set was too large and the approach to building this model was not accurate. See PricingEstimateV2 for newer model